core/src/commonMain/kotlin/com/xebia/functional/xef/llm/ChatWithFunctions.kt

```diff
@@ -31,8 +36,29 @@ suspend fun <A> ChatApi.prompt(
   scope: Conversation,
   serializer: KSerializer<A>,
 ): A =
-  prompt(prompt, scope, chatFunction(serializer.descriptor)) { json ->
-    Json.decodeFromString(serializer, json)
+  prompt(prompt, scope, chatFunctions(listOf(serializer.descriptor))) { call ->
+    Json.decodeFromString(serializer, call.arguments)
+  }
+
+@OptIn(ExperimentalSerializationApi::class)
+@AiDsl
+suspend fun <A> ChatApi.prompt(
+  prompt: Prompt<CreateChatCompletionRequestModel>,
+  scope: Conversation,
+  serializer: KSerializer<A>,
+  descriptors: List<SerialDescriptor>,
+): A =
+  prompt(prompt, scope, chatFunctions(descriptors)) { call ->
+    // adds a `type` field with the call.functionName serial name equivalent to the call arguments
+    val jsonWithDiscriminator = Json.decodeFromString(JsonElement.serializer(), call.arguments)
+    val descriptor =
+      descriptors.firstOrNull { it.serialName.endsWith(call.functionName) }
```

Comment:
This can be an issue because we use simple function names without FQN. Sadly, sending the full package name to Open AI is not a good idea because selecting the proper function makes it more challenging. We bail if the function is not found in the cases, forcing the user to reencode it or provide `@SerialName`.